src/simulation/m_kernel_functions.fpp

```diff
@@ -0,0 +1,501 @@
+            !! @param lag_bub_motion_pos Spatial coordinates of the bubbles
+            !! @param updatedvar Eulerian variable to be updated
+    subroutine s_smoothfunction(nBubs, lag_bub_interface_rad, lag_bub_interface_vel, lag_bub_motion_s, lag_bub_motion_pos, updatedvar)
+
+        integer, intent(in) :: nBubs
+        real(kind(0d0)), dimension(1:lag_nBubs_glb, 1:3, 1:2), intent(in) :: lag_bub_motion_s, lag_bub_motion_pos
+        real(kind(0d0)), dimension(1:lag_nBubs_glb, 1:2), intent(in) :: lag_bub_interface_rad, lag_bub_interface_vel
+        type(vector_field), intent(inout) :: updatedvar
+
+        mapCells = 3    !Number of cells around the bubble where the strength function will have effect (3+1+3)
+        !$acc update device(mapCells)
+
+        smoothfunc:select case(lag_smooth_type)
+        case (1)
+        call s_gaussian(nBubs, lag_bub_interface_rad, lag_bub_interface_vel, lag_bub_motion_s, lag_bub_motion_pos, updatedvar)
+        case (2)
+        call s_deltafunc(nBubs, lag_bub_interface_rad, lag_bub_interface_vel, lag_bub_motion_s, lag_bub_motion_pos, updatedvar)
+        end select smoothfunc
+
+    end subroutine s_smoothfunction
+
+    !> The purpose of this procedure contains the algorithm to use the delta kernel function to map the effect of the bubbles.
+            !!      The effect of the bubbles only affects the cell where the bubble is located.
+    subroutine s_deltafunc(nBubs, lag_bub_interface_rad, lag_bub_interface_vel, lag_bub_motion_s, lag_bub_motion_pos, updatedvar)
+
+        integer, intent(in) :: nBubs
+        real(kind(0d0)), dimension(1:lag_nBubs_glb, 1:3, 1:2), intent(in) :: lag_bub_motion_s, lag_bub_motion_pos
+        real(kind(0d0)), dimension(1:lag_nBubs_glb, 1:2), intent(in) :: lag_bub_interface_rad, lag_bub_interface_vel
+        type(vector_field), intent(inout) :: updatedvar
+
+        integer, dimension(3) :: cell
+        real(kind(0.d0)) :: strength_vel, strength_vol
+
+        real(kind(0.d0)) :: addFun1, addFun2, addFun3
+        real(kind(0.d0)) :: volpart, Vol
+        real(kind(0.d0)), dimension(3) :: s_coord
+        integer :: l
+
+        !$acc parallel loop gang vector default(present) private(l, s_coord, cell)
+        do l = 1, nBubs
+
+            volpart = 4.0d0/3.0d0*pi*lag_bub_interface_rad(l, 2)**3
+            s_coord(1:3) = lag_bub_motion_s(l, 1:3, 2)
+            call s_get_cell(s_coord, cell)
+
+            strength_vol = volpart
+            strength_vel = 4.0d0*pi*lag_bub_interface_rad(l, 2)**2*lag_bub_interface_vel(l, 2)
+
+            if (num_dims == 2) then
+                Vol = dx(cell(1))*dy(cell(2))*lag_charwidth
+                if (cyl_coord) Vol = dx(cell(1))*dy(cell(2))*y_cc(cell(2))*2d0*pi
+            else
+                Vol = dx(cell(1))*dy(cell(2))*dz(cell(3))
+            end if
+
+            !Update void fraction field
+            addFun1 = strength_vol/Vol
+            !$acc atomic update
```

Comment:
calling attention to this line. might need careful attention for CCE+AMD GPU.

Reply:
This atomic function is a strategy I came across online to prevent the possibility of different threads simultaneously overwriting the same cell, which could occur if two bubbles are sufficiently close together.